lib/spack/spack/build_systems/cached_cmake.py

```diff
@@ -132,6 +132,11 @@ def initconfig_compiler_entries(self):
             "endif()\n",
         ]
 
+        # We defined hipcc as top-level compiler for packages when +rocm.
+        # This avoid problems coming from rocm flags being applied to another compiler.
+        if "+rocm" in spec:
+            entries.insert(0, cmake_cache_path("CMAKE_CXX_COMPILER", self.spec["hip"].hipcc))
```

Comment:
I'm curious what goes wrong when this is removed. I'm also not clear what you mean with 

> This avoid problems coming from rocm flags being applied to another compiler.

do you mean this isn't needed for some cases of `spec%compiler` but is needed for others? Or is this because you have HIP-specific compiler flags (e.g. in your `compilers.yaml`)?

Reply:
A simple example of what goes wrong without this is that gcc will not recognize `--rocm-path=...`.

Reply:
Right, but hipcc shouldn't be set as the CMAKE_CXX_COMPILER, but the CMAKE_HIP_COMPILER unless something in spack is accidentally passing a flag to CXX_FLAGS that doesn't belong there.

Reply:
This is required after switching to the native target-based support hip in BLT. This means a HIP capable compiler must be used, but as Adrien says, Spack CI builds things that are +rocm%gcc, which really doesn't make sense since HIP has no notion of host compiler in the way cuda does.

Reply:
Thanks @davidbeckingsale. I remember trying CMAKE_HIP_COMPILER, which is actually what I have in #39613. But the tests there are failing with things like.
```
  >> 128    g++: error: unrecognized command-line option '--rocm-path=/opt/rocm
            -5.4.3'
```
I could re-run the same test here, but I’d like to emphasize that I am only doing what was already there before this PR:
https://github.com/spack/spack/blob/22ad28bb17820eed041731814950531e24d52a30/var/spack/repos/builtin/packages/raja/package.py#L196

The choice here is not to solve everything, trying that got me into too many problems at the same time, but rather to make some progress, incrementally. Hopefully gathering the logic in a single location is seen as a form of progress ;).

Reply:
I'm probably missing something, but wouldn't that mean `CMAKE_CXX_COMPILER=g++` and `CMAKE_HIP_COMPILER=amdclang++`?  Using `hipcc` doesn't seem to make that better so much as more confusing.  If we end up having to leave this and fix it in another issue, I can probably live with it, but it seems pretty unfortunate that all non-hip code is being compiled with an entirely different compiler than spack has been asked to use, rather than just all host code _in hip compilation units_.

Reply:
> This was already in Spack, both below in the CachedCMakePackage but through the HIP_CXX_COMPILER, which I removed, and in the packages, using the CMAKE_CXX_COMPILER

I thought HIP_CXX_COMPILER and CMAKE_CXX_COMPILER were different: the former is used specifically for hip-enabled codes (which might not be all the code in the project).

Apologies if I'm rehashing what's already been stated.

>  g++: error: unrecognized command-line option '--rocm-path=/opt/rocm-5.4.3'

If I understand, you have explicit cflags in your config somewhere. If I'm right, most CMake packages set ROCM_PATH (at which point CMake knows how to invoke each compiler with the proper args).